In [38]:
import argparse
import random
import torch
from torch import nn
import numpy as np
from tqdm import tqdm
import functools
import os
import cProfile
import copy
import time
import sys
import json
import subprocess
from torch import nn
from sklearn.model_selection import train_test_split
import timm
import pandas
import seaborn
import matplotlib.pyplot
import numpy as np
from tqdm import tqdm
from collections import Counter
import math
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap
import numpy as np
import csv
import argparse
import sys
import ast
from sklearn.metrics.pairwise import cosine_similarity, euclidean_distances
from typing import Tuple, Dict



import wandb
from sklearn.metrics.pairwise import cosine_similarity

from torch.utils.data import DataLoader

from src.backbone import get_backbone
from src.modules import CosineLinear
from src.moe_seed import MoE_SEED
from src.data import (
    CILDataManager,
    DILDataManager,
    get_dataset,
    DATASET_MAP,
    make_test_transform_from_args,
    make_train_transform_from_args,
    update_transforms,
)
from src.logging import Logger, WandbLogger, ConsoleLogger, TQDMLogger
from torch.utils.data import Subset

from src.support_functions import check_gpu_memory, shrink_dataset, display_profile, log_gpustat, optimize_args



# Important! Only one dataset at a time

In [5]:
DATASET = "vtab"
CUDA_VISIBLE_DEVICES = "4"

## Get dataset or class features

In [39]:

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False


def setup_logger(args):
    Logger.instance().add_backend(ConsoleLogger())
    if args.wandb_project is not None:
        Logger.instance().add_backend(
            WandbLogger(args.wandb_project, args.wandb_entity, args)
        )



def update_args(args):
    assert args.k >= 1 and args.k <= 12
    args.intralayers = [f"blocks.{11 - i}" for i in range(args.k)]

    args.aug_normalize = bool(args.aug_normalize)

    args.target_size = 224
    
    dataset_T_map = {
        "dil_imagenetr": {"T": 15, "moe_max_experts": 2},
        "limited_domainnet": {"T": 6, "moe_max_experts": 3},
        "vtab": {"T": 5, "moe_max_experts": 3},
        "cddb": {"T": 5, "moe_max_experts": 3},
    }

    if args.dataset in dataset_T_map.keys():
        args.T = dataset_T_map[args.dataset]["T"]
        #args.moe_max_experts = dataset_T_map[args.dataset]["moe_max_experts"] #immer 5!
        print(f"Dataset {args.dataset} has T={args.T} and moe_max_experts={args.moe_max_experts}")

    return args




def calculate_similarity(feature1, feature2, metric='cosine'):
    """
    Calculates the similarity between two feature vectors.

    Args:
        feature1 (np.ndarray): First feature vector.
        feature2 (np.ndarray): Second feature vector.
        metric (str): The similarity metric to use ('cosine' or 'euclidean').

    Returns:
        float: The similarity score.
    """
    if metric == 'cosine':
        return cosine_similarity(feature1.reshape(1, -1), feature2.reshape(1, -1))[0][0]
    elif metric == 'euclidean':
        return -np.linalg.norm(feature1 - feature2) # Negative for consistency (higher value = more similar)
    else:
        raise ValueError(f"Unsupported similarity metric: {metric}")

def calculate_intra_class_similarity(features, labels, similarity_metric='cosine'):
    """
    Calculates the average intra-class similarity for a dataset.

    Args:
        features (np.ndarray): Array of feature vectors.
        labels (np.ndarray): Array of corresponding labels.
        similarity_metric (str): The similarity metric to use ('cosine' or 'euclidean').

    Returns:
        tuple: (dict, float) - A dictionary of per-class intra-class similarities
               and the overall average intra-class similarity.
    """
    intra_class_similarities = {}
    unique_labels = np.unique(labels)
    bar = tqdm(enumerate(unique_labels), desc="Calculating Intra-Class Similarity", total=len(unique_labels))
    for i, label in bar:
        class_features = features[labels == label]
        similarities = []
        for i in range(len(class_features)):
            for j in range(i + 1, len(class_features)):
                similarity = calculate_similarity(class_features[i], class_features[j], similarity_metric)
                similarities.append(similarity)
        if similarities:
            intra_class_similarities[label] = np.mean(similarities)
        else:
            intra_class_similarities[label] = 0.0

    overall_intra_class_similarity = np.mean(list(intra_class_similarities.values())) if intra_class_similarities else 0.0
    return intra_class_similarities, overall_intra_class_similarity

# new for intra... last chapter
def calculate_similarity(feature1: np.ndarray, feature2: np.ndarray, metric: str = 'cosine') -> float:
    """
    Calculates the similarity between two feature vectors.

    Args:
        feature1 (np.ndarray): First feature vector.
        feature2 (np.ndarray): Second feature vector.
        metric (str): The similarity metric to use ('cosine' or 'euclidean').

    Returns:
        float: The similarity score.
    """
    if metric == 'cosine':
        # Ensure inputs are 2D for scikit-learn's cosine_similarity
        return cosine_similarity(feature1.reshape(1, -1), feature2.reshape(1, -1))[0][0]
    elif metric == 'euclidean':
        # euclidean_distances returns distances, so we negate for similarity.
        # It's more efficient to calculate norm directly for two vectors.
        return -np.linalg.norm(feature1 - feature2)
    else:
        raise ValueError(f"Unsupported similarity metric: {metric}")

def calculate_intra_class_similarity_optimized(features: np.ndarray, labels: np.ndarray, similarity_metric: str = 'cosine') -> Tuple[dict, float]:
    """
    Calculates the average intra-class similarity for a dataset efficiently.

    Args:
        features (np.ndarray): Array of feature vectors (n_samples, n_features).
        labels (np.ndarray): Array of corresponding labels (n_samples,).
        similarity_metric (str): The similarity metric to use ('cosine' or 'euclidean').

    Returns:
        tuple: (dict, float) - A dictionary of per-class intra-class similarities
               and the overall average intra-class similarity.
    """
    intra_class_similarities = {}
    unique_labels = np.unique(labels)

    # Choose the appropriate pairwise function based on the metric
    if similarity_metric == 'cosine':
        pairwise_func = cosine_similarity
    elif similarity_metric == 'euclidean':
        # For Euclidean distance, we'll use euclidean_distances and then negate it
        pairwise_func = euclidean_distances
    else:
        raise ValueError(f"Unsupported similarity metric: {similarity_metric}")

    bar = tqdm(unique_labels, desc="Calculating Intra-Class Similarity", unit="class")
    for label in bar:
        class_features = features[labels == label]
        n_samples_in_class = class_features.shape[0]

        if n_samples_in_class < 2:
            # Cannot calculate pairwise similarity for 0 or 1 sample
            intra_class_similarities[label] = 0.0
            continue

        # Calculate all-to-all pairwise similarities within the class
        # This is where the main efficiency gain comes from
        pairwise_sims_or_dists = pairwise_func(class_features)

        # Extract only the upper triangle (excluding diagonal) to avoid duplicates and self-similarity
        # np.triu_indices_from gets indices for upper triangle
        upper_tri_indices = np.triu_indices_from(pairwise_sims_or_dists, k=1)
        
        similarities = pairwise_sims_or_dists[upper_tri_indices]

        # If Euclidean, negate the distances to represent similarity
        if similarity_metric == 'euclidean':
            similarities = -similarities
        
        if similarities.size > 0:
            intra_class_similarities[label] = np.mean(similarities)
        else:
            intra_class_similarities[label] = 0.0 # Should not happen if n_samples_in_class >= 2

    overall_intra_class_similarity = np.mean(list(intra_class_similarities.values())) if intra_class_similarities else 0.0
    
    return intra_class_similarities, overall_intra_class_similarity


def calculate_inter_class_similarity(features, labels, similarity_metric='cosine'):
    """
    Calculates the average inter-class similarity for a dataset.

    Args:
        features (np.ndarray): Array of feature vectors.
        labels (np.ndarray): Array of corresponding labels.
        similarity_metric (str): The similarity metric to use ('cosine' or 'euclidean').

    Returns:
        float: The overall average inter-class similarity.
    """
    inter_class_similarities = []
    unique_labels = np.unique(labels)
    bar = tqdm(enumerate(unique_labels), desc="Calculating Inter-Class Similarity", total=len(unique_labels))
    # Iterate through all pairs of classes
    for i, _ in bar:
        for j in range(i + 1, len(unique_labels)):
            label1 = unique_labels[i]
            label2 = unique_labels[j]
            features_class1 = features[labels == label1]
            features_class2 = features[labels == label2]
            for feat1 in features_class1:
                for feat2 in features_class2:
                    similarity = calculate_similarity(feat1, feat2, similarity_metric)
                    inter_class_similarities.append(similarity)

    overall_inter_class_similarity = np.mean(inter_class_similarities) if inter_class_similarities else 0.0
    return overall_inter_class_similarity

def calculate_inter_class_similarity_vectorized(features, labels, similarity_metric='cosine'):
    """
    Calculates the average inter-class similarity for a dataset using vectorization.
    """
    inter_class_similarities = []
    unique_labels = np.unique(labels)
    n_classes = len(unique_labels)
    bar = tqdm(range(n_classes), desc="Calculating Inter-Class Similarity", total=n_classes)

    for i in bar:
        for j in range(i + 1, n_classes):
            label1 = unique_labels[i]
            label2 = unique_labels[j]
            features_class1 = features[labels == label1]
            features_class2 = features[labels == label2]

            if similarity_metric == 'cosine':
                similarity_matrix = cosine_similarity(features_class1, features_class2)
                inter_class_similarities.extend(similarity_matrix.flatten())
            elif similarity_metric == 'euclidean':
                # Calculate pairwise Euclidean distances and negate for consistency
                distances = np.linalg.norm(features_class1[:, np.newaxis, :] - features_class2[np.newaxis, :, :], axis=2)
                inter_class_similarities.extend((-distances).flatten())
            else:
                raise ValueError(f"Unsupported similarity metric: {similarity_metric}")

    overall_inter_class_similarity = np.mean(inter_class_similarities) if inter_class_similarities else 0.0
    return overall_inter_class_similarity

# Expert learned a set of classes and we want to calculate the similarity to all other classes
def calculate_selective_inter_class_similarity(features, labels, target_classes, similarity_metric='cosine'):
    """
    Calculates the average inter-class similarity between a target list of classes
    and all other classes not in the target list, using vectorization.

    Args:
        features (np.ndarray): Array of feature vectors.
        labels (np.ndarray): Array of corresponding labels.
        target_classes (list): List of class labels for which to compute similarity
                               to all other classes (excluding those in this list).
        similarity_metric (str, optional): Metric to use for similarity calculation.
                                           'cosine' or 'euclidean' are supported.
                                           Defaults to 'cosine'.

    Returns:
        float: The average inter-class similarity between the target classes
               and the other classes. Returns 0.0 if no such pairs exist.
    """
    inter_class_similarities = []
    unique_labels = np.unique(labels)
    target_classes = set(target_classes)  # Convert to set for faster lookups
    other_classes = [label for label in unique_labels if label not in target_classes]

    bar = tqdm(target_classes, desc="Calculating Selective Inter-Class Similarity", total=len(target_classes))

    for label1 in bar:
        features_class1 = features[labels == label1]
        for label2 in other_classes:
            features_class2 = features[labels == label2]

            if similarity_metric == 'cosine':
                similarity_matrix = cosine_similarity(features_class1, features_class2)
                inter_class_similarities.extend(similarity_matrix.flatten())
            elif similarity_metric == 'euclidean':
                # Calculate pairwise Euclidean distances and negate for consistency
                distances = np.linalg.norm(features_class1[:, np.newaxis, :] - features_class2[np.newaxis, :, :], axis=2)
                inter_class_similarities.extend((-distances).flatten())
            else:
                raise ValueError(f"Unsupported similarity metric: {similarity_metric}")

    overall_inter_class_similarity = np.mean(inter_class_similarities) if inter_class_similarities else 0.0
    return overall_inter_class_similarity


def calculate_entropy(labels):
    """
    Calculates the entropy of a list or NumPy array of labels.

    Args:
        labels (list or np.ndarray): A list or array of labels.

    Returns:
        float: The entropy of the labels.
    """

    label_counts = Counter(labels)
    total_samples = len(labels)
    entropy = 0.0

    for count in label_counts.values():
        probability = count / total_samples
        entropy -= probability * math.log2(probability)

    return entropy

def calculate_feature_variance(features):
    """
    Calculates the variance of the features.

    Args:
        features (np.ndarray): Array of feature vectors.

    Returns:
        float: The variance of the features.
    """
    return np.var(features, axis=0).mean()  # Mean variance across all features

def visualize_csv_with_adjusted_size(csv_filepath, output_filepath="heatmap_adjusted.png"):
    try:
        df = pd.read_csv(csv_filepath, index_col=0)
    except FileNotFoundError:
        print(f"Fehler: Datei '{csv_filepath}' nicht gefunden.")
        return

    numeric_cols = df.select_dtypes(include=np.number).columns

    if not numeric_cols.empty:
        df_normalized = df[numeric_cols].apply(lambda x: (x - x.min()) / (x.max() - x.min()), axis=0)
        cmap = LinearSegmentedColormap.from_list("mycmap", ["white", "lightblue", "darkblue"])

        # Erhöhe die Figurengröße, um die Kästchen größer zu machen
        plt.figure(figsize=(len(numeric_cols) * 2, len(df) * 1))

        sns.heatmap(df_normalized, annot=False, cmap=cmap, cbar=True, yticklabels=True)
        plt.title("Farbliche Visualisierung der Datenspalten", fontsize=12) # Kleinere Schriftgröße für den Titel
        plt.xlabel("Numerische Spalten", fontsize=10) # Kleinere Schriftgröße für die X-Achse
        plt.ylabel("Datensätze", fontsize=10) # Kleinere Schriftgröße für die Y-Achse
        plt.xticks(rotation=45, ha="right", fontsize=8) # Kleinere Schriftgröße für die X-Achsenbeschriftungen
        plt.yticks(fontsize=8) # Kleinere Schriftgröße für die Y-Achsenbeschriftungen
        plt.tight_layout()
        plt.savefig(output_filepath)
        print(f"Heatmap mit angepasster Größe und Schrift gespeichert als '{output_filepath}'.")
        plt.close()
    else:
        print("Keine numerischen Spalten zum Visualisieren gefunden.")


def get_all_dataset_metrics(train_features, train_labels, dataset_name):
    values = [None] * 6
    values[0] = dataset_name
    
    label_entropy = calculate_entropy(train_labels)
    print(f"Label Entropy: {label_entropy:.4f}")
    values[1] = label_entropy
    feature_entropy = calculate_entropy(train_features.flatten())
    print(f"Feature Entropy: {feature_entropy:.4f}")
    values[2] = feature_entropy

    # 4. Calculate intra-class similarity on the training set
    intra_class_similarities, overall_intra_similarity = calculate_intra_class_similarity(train_features, train_labels) # Use train_labels here
    print(f"Intra-Class Similarities per class: {intra_class_similarities}")
    print(f"Overall Intra-Class Similarity: {overall_intra_similarity:.4f}")
    values[5] = overall_intra_similarity
    
    # 5. Calculate inter-class similarity on the training set
    overall_inter_similarity = calculate_inter_class_similarity_vectorized(train_features, train_labels) # Use train_labels here
    print(f"Overall Inter-Class Similarity: {overall_inter_similarity:.4f}")
    values[4] = overall_inter_similarity

    # 6. Calculate feature variance
    feature_variance = calculate_feature_variance(train_features)
    print(f"Feature Variance: {feature_variance:.4f}")
    values[3] = feature_variance

    return values


def main(args):
    # get dataset and augmentations
    train_transform = make_train_transform_from_args(args)
    test_transform = make_test_transform_from_args(args)
    train_base_dataset, test_base_dataset = get_dataset(
        args.dataset, path=args.data_root
    )
    update_transforms(test_base_dataset, transform=test_transform)


    # get datamanager based on ds
    data_manager = None
    if DILDataManager.is_dil(str(train_base_dataset)):
        print("DIL")
        data_manager = DILDataManager(
            train_base_dataset,
            test_base_dataset,
        )
    else:
        print("CIL")
        data_manager = CILDataManager(
            train_base_dataset,
            test_base_dataset,
            T=args.T,
            num_first_task=None if args.dataset != "cars" else 16,
            shuffle=True,
            seed=args.seed,
        )


    feature_extractor = timm.create_model(args.backbone, pretrained=True).to(args.device)
    feature_extractor.head = nn.Identity()
    feature_extractor.eval()
    train_features = []
    train_labels = []
    bar = tqdm(enumerate(data_manager), desc="Extracting Features", total=len(data_manager))
    for i, (train_dataset, _) in bar:
        train_dataset.transform = train_transform
        train_loader = DataLoader(train_dataset, batch_size=args.batch_size, shuffle=False, num_workers=2, pin_memory=True)
        for images, labels in train_loader:  # Iterate through all batches in the loader
            images = images.to(args.device)
            features = feature_extractor(images)
            train_features.append(features.cpu().detach().numpy())
            train_labels.append(labels.cpu().detach().numpy())

    del feature_extractor

    train_features = np.concatenate(train_features, axis=0)
    train_labels = np.concatenate(train_labels, axis=0)
    print("Features shape:", train_features.shape)

    return train_features, train_labels   

    #values = get_all_dataset_metrics(train_features, train_labels, args.dataset)
    
"""
    # Saving values
    save_path = "local_data/dataset_metrics.csv"
    with open(save_path, 'a', newline='') as outfile:
        writer = csv.writer(outfile)
        writer.writerow(values)
"""


def main2_classes_per_task_and_seed(args):
    # get dataset and augmentations
    train_transform = make_train_transform_from_args(args)
    test_transform = make_test_transform_from_args(args)
    train_base_dataset, test_base_dataset = get_dataset(
        args.dataset, path=args.data_root
    )
    update_transforms(test_base_dataset, transform=test_transform)


    # get datamanager based on ds
    data_manager = None
    if DILDataManager.is_dil(str(train_base_dataset)):
        print("DIL")
        data_manager = DILDataManager(
            train_base_dataset,
            test_base_dataset,
        )
    else:
        print("CIL")
        data_manager = CILDataManager(
            train_base_dataset,
            test_base_dataset,
            T=args.T,
            num_first_task=None if args.dataset != "cars" else 16,
            shuffle=True,
            seed=args.seed,
        )
        class_order = data_manager.class_order

    csv_rows = []
    bar = tqdm(enumerate(data_manager), desc="Accumulating classes per task", total=len(data_manager))
    for t, (train_dataset, _) in bar:
        task_labels_set = np.array([], dtype=np.int_)

        train_dataset.transform = train_transform
        train_loader = DataLoader(train_dataset, batch_size=args.batch_size, shuffle=False, num_workers=2, pin_memory=True)
        for _, labels in train_loader:  # Iterate through all batches in the loader
            batch_labels = labels.cpu().detach().numpy()
            task_labels_set = np.concatenate((task_labels_set, batch_labels))
        # Get unique labels
        _, idx = np.unique(task_labels_set, return_index=True)
        task_labels_set = task_labels_set[np.sort(idx)]
        # label mapping
        for i, e in enumerate(task_labels_set):
            task_labels_set[i] = class_order[e]

        for i in task_labels_set:
            row = {
                "dataset": args.dataset,
                "task": t,
                "class": i,
                "seed": args.seed
            }
            csv_rows.append(row)




    # Save the results to a CSV file
    save_path = f"local_data/classes_per_task/CpT_{args.dataset}_{args.seed}.csv"
    os.makedirs(os.path.dirname(save_path), exist_ok=True)
    with open(save_path, 'w', newline='') as outfile:
        writer = csv.DictWriter(outfile, fieldnames=["dataset", "task", "class", "seed"])
        writer.writeheader()
        writer.writerows(csv_rows)
    print(f"Classes per task saved in {save_path}")

In [9]:
os.environ['CUDA_VISIBLE_DEVICES'] = "1"
args = parse_arguments()


seeds = [2000, 2001, 2002, 2003, 2004]
datasets = ["imagenetr", "cifar100", "cub", "imageneta", "cars", "omnibenchmark"]


for seed in seeds:
    for dataset in datasets:
        args.seed = seed
        args.dataset = dataset
        args.device = "cpu"
        args = update_args(args)

        set_seed(args.seed)

        main2_classes_per_task_and_seed(args)

CIL


Iterate over train dataset: 24000it [00:00, 663822.00it/s]
Iterate over test dataset: 6000it [00:00, 1166272.31it/s]
Accumulating classes per task: 100%|██████████| 10/10 [02:04<00:00, 12.46s/it]


Classes per task saved in local_data/classes_per_task/CpT_imagenetr_2000.csv
Files already downloaded and verified
Files already downloaded and verified
CIL


Iterate over train dataset: 50000it [00:02, 19682.41it/s]
Iterate over test dataset: 10000it [00:14, 691.54it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:27<00:00,  8.77s/it]


Classes per task saved in local_data/classes_per_task/CpT_cifar100_2000.csv
CIL


Iterate over train dataset: 9430it [00:00, 1283040.41it/s]
Iterate over test dataset: 2358it [00:00, 1365668.16it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:44<00:00,  4.50s/it]


Classes per task saved in local_data/classes_per_task/CpT_cub_2000.csv
CIL


Iterate over train dataset: 5981it [00:00, 1215177.88it/s]
Iterate over test dataset: 1519it [00:00, 1339039.04it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:30<00:00,  3.04s/it]


Classes per task saved in local_data/classes_per_task/CpT_imageneta_2000.csv
CIL


Iterate over train dataset: 8144it [00:00, 1299688.45it/s]
Iterate over test dataset: 8041it [00:00, 1578507.84it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:53<00:00,  5.31s/it]


Classes per task saved in local_data/classes_per_task/CpT_cars_2000.csv
CIL


Iterate over train dataset: 89697it [00:00, 1184874.10it/s]
Iterate over test dataset: 5985it [00:00, 1477945.80it/s]
Accumulating classes per task: 100%|██████████| 10/10 [10:11<00:00, 61.12s/it]


Classes per task saved in local_data/classes_per_task/CpT_omnibenchmark_2000.csv
CIL


Iterate over train dataset: 24000it [00:00, 1165691.58it/s]
Iterate over test dataset: 6000it [00:00, 1512641.94it/s]
Accumulating classes per task: 100%|██████████| 10/10 [02:06<00:00, 12.66s/it]


Classes per task saved in local_data/classes_per_task/CpT_imagenetr_2001.csv
Files already downloaded and verified
Files already downloaded and verified
CIL


Iterate over train dataset: 50000it [00:03, 16512.66it/s]
Iterate over test dataset: 10000it [00:14, 707.31it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:28<00:00,  8.86s/it]


Classes per task saved in local_data/classes_per_task/CpT_cifar100_2001.csv
CIL


Iterate over train dataset: 9430it [00:00, 444392.74it/s]
Iterate over test dataset: 2358it [00:00, 1123244.61it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:50<00:00,  5.08s/it]


Classes per task saved in local_data/classes_per_task/CpT_cub_2001.csv
CIL


Iterate over train dataset: 5981it [00:00, 1170062.14it/s]
Iterate over test dataset: 1519it [00:00, 1314723.02it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:30<00:00,  3.07s/it]


Classes per task saved in local_data/classes_per_task/CpT_imageneta_2001.csv
CIL


Iterate over train dataset: 8144it [00:00, 666232.60it/s]
Iterate over test dataset: 8041it [00:00, 729298.27it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:55<00:00,  5.59s/it]


Classes per task saved in local_data/classes_per_task/CpT_cars_2001.csv
CIL


Iterate over train dataset: 89697it [00:00, 1280244.49it/s]
Iterate over test dataset: 5985it [00:00, 1547365.43it/s]
Accumulating classes per task: 100%|██████████| 10/10 [10:39<00:00, 63.91s/it]


Classes per task saved in local_data/classes_per_task/CpT_omnibenchmark_2001.csv
CIL


Iterate over train dataset: 24000it [00:00, 683199.49it/s]
Iterate over test dataset: 6000it [00:00, 1235496.29it/s]
Accumulating classes per task: 100%|██████████| 10/10 [02:09<00:00, 12.90s/it]


Classes per task saved in local_data/classes_per_task/CpT_imagenetr_2002.csv
Files already downloaded and verified
Files already downloaded and verified
CIL


Iterate over train dataset: 50000it [00:03, 14157.52it/s]
Iterate over test dataset: 10000it [00:17, 575.21it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:28<00:00,  8.80s/it]


Classes per task saved in local_data/classes_per_task/CpT_cifar100_2002.csv
CIL


Iterate over train dataset: 9430it [00:00, 1214489.72it/s]
Iterate over test dataset: 2358it [00:00, 1379189.63it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:44<00:00,  4.44s/it]


Classes per task saved in local_data/classes_per_task/CpT_cub_2002.csv
CIL


Iterate over train dataset: 5981it [00:00, 1221092.88it/s]
Iterate over test dataset: 1519it [00:00, 1580145.78it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:26<00:00,  2.63s/it]


Classes per task saved in local_data/classes_per_task/CpT_imageneta_2002.csv
CIL


Iterate over train dataset: 8144it [00:00, 1678463.55it/s]
Iterate over test dataset: 8041it [00:00, 1429084.68it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:45<00:00,  4.59s/it]


Classes per task saved in local_data/classes_per_task/CpT_cars_2002.csv
CIL


Iterate over train dataset: 89697it [00:00, 1397363.94it/s]
Iterate over test dataset: 5985it [00:00, 1667413.45it/s]
Accumulating classes per task: 100%|██████████| 10/10 [09:10<00:00, 55.05s/it]


Classes per task saved in local_data/classes_per_task/CpT_omnibenchmark_2002.csv
CIL


Iterate over train dataset: 24000it [00:00, 1325266.87it/s]
Iterate over test dataset: 6000it [00:00, 1877766.30it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:45<00:00, 10.51s/it]


Classes per task saved in local_data/classes_per_task/CpT_imagenetr_2003.csv
Files already downloaded and verified
Files already downloaded and verified
CIL


Iterate over train dataset: 50000it [00:02, 22062.05it/s]
Iterate over test dataset: 10000it [00:12, 798.67it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:20<00:00,  8.06s/it]


Classes per task saved in local_data/classes_per_task/CpT_cifar100_2003.csv
CIL


Iterate over train dataset: 9430it [00:00, 1486145.89it/s]
Iterate over test dataset: 2358it [00:00, 1757315.00it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:42<00:00,  4.28s/it]


Classes per task saved in local_data/classes_per_task/CpT_cub_2003.csv
CIL


Iterate over train dataset: 5981it [00:00, 820371.24it/s]
Iterate over test dataset: 1519it [00:00, 596605.28it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:26<00:00,  2.65s/it]


Classes per task saved in local_data/classes_per_task/CpT_imageneta_2003.csv
CIL


Iterate over train dataset: 8144it [00:00, 1386411.71it/s]
Iterate over test dataset: 8041it [00:00, 1708098.17it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:47<00:00,  4.73s/it]


Classes per task saved in local_data/classes_per_task/CpT_cars_2003.csv
CIL


Iterate over train dataset: 89697it [00:00, 1387080.61it/s]
Iterate over test dataset: 5985it [00:00, 1015038.19it/s]
Accumulating classes per task: 100%|██████████| 10/10 [08:19<00:00, 49.99s/it]


Classes per task saved in local_data/classes_per_task/CpT_omnibenchmark_2003.csv
CIL


Iterate over train dataset: 24000it [00:00, 1209895.38it/s]
Iterate over test dataset: 6000it [00:00, 1847168.53it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:50<00:00, 11.08s/it]


Classes per task saved in local_data/classes_per_task/CpT_imagenetr_2004.csv
Files already downloaded and verified
Files already downloaded and verified
CIL


Iterate over train dataset: 50000it [00:02, 17587.85it/s]
Iterate over test dataset: 10000it [00:15, 665.26it/s]
Accumulating classes per task: 100%|██████████| 10/10 [01:28<00:00,  8.83s/it]


Classes per task saved in local_data/classes_per_task/CpT_cifar100_2004.csv
CIL


Iterate over train dataset: 9430it [00:00, 875050.59it/s]
Iterate over test dataset: 2358it [00:00, 1012092.59it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:47<00:00,  4.78s/it]


Classes per task saved in local_data/classes_per_task/CpT_cub_2004.csv
CIL


Iterate over train dataset: 5981it [00:00, 1262957.87it/s]
Iterate over test dataset: 1519it [00:00, 1050651.02it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:30<00:00,  3.10s/it]


Classes per task saved in local_data/classes_per_task/CpT_imageneta_2004.csv
CIL


Iterate over train dataset: 8144it [00:00, 1320693.31it/s]
Iterate over test dataset: 8041it [00:00, 1280522.38it/s]
Accumulating classes per task: 100%|██████████| 10/10 [00:56<00:00,  5.65s/it]


Classes per task saved in local_data/classes_per_task/CpT_cars_2004.csv
CIL


Iterate over train dataset: 89697it [00:00, 943193.95it/s]
Iterate over test dataset: 5985it [00:00, 1110994.00it/s]
Accumulating classes per task: 100%|██████████| 10/10 [09:02<00:00, 54.28s/it]

Classes per task saved in local_data/classes_per_task/CpT_omnibenchmark_2004.csv


In [3]:
def parse_arguments():
    parser = argparse.ArgumentParser(description="Your script description here")

    # Define your arguments as before
    parser.add_argument('--batch_size', type=int, default=32, help='Batch size for training')
    parser.add_argument('--lr', type=float, default=0.001, help='Learning rate')
    parser.add_argument('--weight_decay', type=float, default=0.0, help='Weight decay')
    parser.add_argument('--early_stopping', type=int, default=10, help='Patience for early stopping')
    parser.add_argument('--dataset', type=str, default='cifar100',
                        choices=['cifar100', 'imagenetr', 'imageneta', 'vtab', 'cars', 'cub',
                                 'omnibenchmark', 'dil_imagenetr', 'cddb', 'limited_domainnet'],
                        help='Dataset to use')
    parser.add_argument('--T', type=int, default=10, help='Number of timesteps')
    parser.add_argument('--backbone', type=str, default='vit_base_patch16_224',
                        choices=['vit_base_patch16_224', 'vit_base_patch16_224_in21k'],
                        help='Backbone architecture')
    parser.add_argument('--finetune_method', type=str, default='none',
                        choices=['none', 'adapter', 'ssf', 'vpt'],
                        help='Finetuning method')
    parser.add_argument('--finetune_epochs', type=int, default=20, help='Number of finetuning epochs')
    parser.add_argument('--k', type=int, default=5, help='Number of nearest neighbors')
    parser.add_argument('--device', type=str, default='cuda' if 'cuda' in sys.modules else 'cpu',
                        help='Device to use (cuda or cpu)')
    parser.add_argument('--seed', type=int, default=2001, help='Random seed')
    parser.add_argument('--data_root', type=str, default='./local_data', help='Root directory for datasets')
    parser.add_argument('--moe_max_experts', type=int, default=4, help='Maximum number of experts for MoE')
    parser.add_argument('--reduce_dataset', type=float, default=1.0, help='Fraction of dataset to use')
    parser.add_argument('--gmms', type=int, default=8, help='Number of GMM components')
    parser.add_argument('--use_multivariate', action='store_true', help='Use multivariate Gaussian')
    parser.add_argument('--selection_method', type=str, default='random',
                        choices=['random', 'around', 'eucld_dist', 'inv_eucld_dist', 'kl_div',
                                 'inv_kl_div', 'ws_div', 'inv_ws_div'],
                        help='Selection method')
    parser.add_argument('--kd', action='store_true', help='Use knowledge distillation')
    parser.add_argument('--kd_alpha', type=float, default=0.5, help='Alpha for knowledge distillation loss')
    parser.add_argument('--log_gpustat', action='store_true', help='Log GPU statistics')
    parser.add_argument('--sweep_logging', action='store_true', help='Enable Weights & Biases sweep logging')
    parser.add_argument('--exit_after_T', action='store_true', help='Exit after T timesteps')
    parser.add_argument('--selection_criterion', type=int, default=0, choices=[0, 1, 2],
                        help='Selection criterion')
    parser.add_argument('--tau', type=float, default=0.1, help='Temperature parameter')
    parser.add_argument('--exit_after_acc', type=float, default=0.0, help='Exit after reaching this accuracy')
    parser.add_argument('--trash_var', type=float, default=0.0, help='Trash variable (for testing)')
    parser.add_argument('--use_adamw', action='store_true', help='Use AdamW optimizer')
    parser.add_argument('--use_cosine_annealing', action='store_true', help='Use cosine annealing scheduler')
    parser.add_argument('--aug_resize_crop_min', type=float, default=0.8, help='Min scale for random resize crop')
    parser.add_argument('--aug_resize_crop_max', type=float, default=1.0, help='Max scale for random resize crop')
    parser.add_argument('--aug_random_rotation_degree', type=float, default=0.0, help='Degree for random rotation')
    parser.add_argument('--aug_brightness_jitter', type=float, default=0.0, help='Brightness jitter')
    parser.add_argument('--aug_contrast_jitter', type=float, default=0.0, help='Contrast jitter')
    parser.add_argument('--aug_saturation_jitter', type=float, default=0.0, help='Saturation jitter')
    parser.add_argument('--aug_hue_jitter', type=float, default=0.0, help='Hue jitter')
    parser.add_argument('--aug_normalize', action='store_true', help='Normalize input images')
    parser.add_argument('--wandb_project', type=str, default='your_project_name', help='WandB project name')
    parser.add_argument('--wandb_entity', type=str, default='your_entity_name', help='WandB entity name')

    if '__file__' in globals():  # Check if running as a script
        args = parser.parse_args()
    else:  # Running in a Jupyter Notebook
        args = parser.parse_args(args=[]) # Pass an empty list to avoid errors
        # You can set default values here or handle arguments differently in the notebook

    return args

In [8]:
os.environ['CUDA_VISIBLE_DEVICES'] = CUDA_VISIBLE_DEVICES
args = parse_arguments()
args.dataset = DATASET
args = update_args(args)
set_seed(args.seed)


features, labels = main(args)

CIL


Iterate over train dataset: 1796it [00:00, 1226069.33it/s]
Iterate over test dataset: 8619it [00:00, 1487438.54it/s]
Extracting Features: 100%|██████████| 5/5 [30:13<00:00, 362.69s/it]

Features shape: (1796, 768)


## Saving all features into file 

In [ ]:
dataset_name = DATASET
labels_list = labels
features_list = features

# Stelle sicher, dass features_list eine Python-Liste ist
if not isinstance(features_list, list):
    features_list = features_list.tolist()  # Konvertiere zu Liste, falls es ein NumPy Array ist

# Annahme: Jede innere Liste hat die gleiche Länge und entspricht einer Zeile
# und die Elemente der inneren Liste sollen separate Spalten werden.

# Erstelle Spaltennamen für die Features
num_features = len(features_list[0]) if features_list else 0
feature_columns = [f'feature_{i}' for i in range(num_features)]

# Erstelle ein Dictionary für den DataFrame
data = {'dataset': [dataset_name] * len(labels_list),
        'label': [item[0] if isinstance(item, list) else item for item in labels_list]} # Annahme: Label ist das erste Element der inneren Liste
for i, col in enumerate(feature_columns):
    data[col] = [item[i] if isinstance(item, list) and len(item) > i else None for item in features_list]

df = pd.DataFrame(data)

# Speichern als CSV-Datei
csv_filename = f"./local_data/{dataset_name}_class_features.csv"
df.to_csv(csv_filename, index=False)
print(f"Daten erfolgreich als '{csv_filename}' gespeichert.")


Daten erfolgreich als './local_data/vtab_class_features.csv' gespeichert (korrigiert).


## Rename class labels
- Die Labels werden ranomised nach seed, werden im Modell aber immer normal durchnummeriert. Das führt aber zu einem missmatch mit den erstellten Features, weil die eine ganz bestimmte random Nummerierung haben, aber 0-99 gelabelt werden.  
- Die Labelreihenfolge wurde bei DIL nicht gelogged. Ich beschränke mich auf CIL!
- Features werden neu erstellt. Mit shuffle=False. Mit Logs (class order =>) überprüfen

In [ ]:
def get_class_order_from_logs(run_id):
    """Lädt die Log-Datei eines WandB-Runs herunter und gibt die Zeile mit "class order => ..." zurück.

    Args:
        run_id (str): Die ID des WandB-Runs.

    Returns:
        str: Die Zeile, die mit "class order => ..." beginnt, oder None, falls nicht gefunden.
    """
    api = wandb.Api()
    try:
        run = api.run(run_id)  # Ersetze mit deiner Entity und deinem Projekt
        if run.files:
            for f in run.files():
                if f.name == 'output.log':  # Der typische Name für die stdout/stderr Logs
                    log_content = f.download(root='.', replace=True, exist_ok=True)
                    if log_content:
                        with open(log_content.name, 'r') as log_file:
                            for line in log_file:
                                if line.startswith("class_order =>"):
                                    print(f"Gefundene Zeile: {line.strip()}")
                                    os.remove(log_content.name) # Optional: Lösche die temporäre Log-Datei
                                    return line.strip()                                
                        os.remove(log_content.name) # Optional: Lösche die temporäre Log-Datei
                        print(f"Keine Zeile mit 'class order => ...' in der Log-Datei für Run {run_id} gefunden.")
                        return None
        print(f"Keine 'output.log'-Datei für Run {run_id} gefunden.")
        return None
    except wandb.CommError as e:
        print(f"Fehler beim Abrufen des Runs {run_id}: {e}")
        return None

# Beispielhafte Verwendung
run_path = "belaschindler-university-hamburg/0schindl-LayUp_sweeps_question1/runs/crnqagtv"  # Ersetze mit der tatsächlichen Run-ID
class_order_line = get_class_order_from_logs(run_path)

if class_order_line:
    print(f"Die Zeile mit 'class order => ...' ist: {class_order_line}")
else:
    print("Die Zeile mit 'class order => ...' wurde nicht gefunden.")

class_order_line = np.array(class_order_line.lstrip("class_order =>"))


Gefundene Zeile: class_order => [164, 106, 3, 122, 90, 148, 76, 193, 179, 177, 142, 192, 6, 43, 131, 30, 104, 20, 160, 115, 80, 69, 119, 55, 99, 78, 172, 185, 145, 8, 85, 47, 154, 13, 88, 195, 19, 178, 132, 51, 81, 70, 170, 87, 184, 113, 14, 7, 125, 117, 139, 79, 50, 107, 114, 97, 92, 174, 77, 41, 54, 150, 108, 35, 156, 149, 56, 130, 110, 40, 135, 196, 17, 91, 10, 173, 37, 22, 32, 100, 189, 12, 116, 66, 199, 71, 23, 161, 136, 94, 36, 158, 52, 67, 27, 187, 153, 63, 98, 33, 45, 74, 18, 93, 140, 59, 190, 165, 83, 15, 191, 5, 157, 123, 57, 138, 144, 53, 95, 166, 65, 49, 38, 25, 58, 11, 163, 75, 152, 89, 96, 159, 64, 34, 167, 168, 2, 176, 103, 102, 124, 29, 28, 73, 24, 180, 162, 84, 31, 197, 137, 126, 182, 133, 111, 61, 46, 155, 181, 188, 127, 21, 82, 169, 9, 143, 48, 109, 120, 183, 194, 146, 62, 112, 118, 129, 151, 175, 141, 86, 105, 72, 147, 39, 198, 121, 26, 44, 60, 171, 68, 42, 128, 1, 101, 186, 16, 134, 4, 0]
Die Zeile mit 'class order => ...' ist: class_order => [164, 106, 3, 122, 90,

## Load data from file

In [6]:
csv_filename = f"./local_data/class_features/class_features_{DATASET}.csv"

# Lese die CSV-Datei ein
df_loaded = pd.read_csv(csv_filename)

# Extrahiere die Labels und Features
loaded_labels = df_loaded['label'].tolist()

# Wenn die Features als separate Spalten gespeichert wurden (Szenario 1):
if 'feature_0' in df_loaded.columns:
    loaded_features = []
    i = 0
    while f'feature_{i}' in df_loaded.columns:
        loaded_features.append(df_loaded[f'feature_{i}'].tolist())
        i += 1
    # Transponiere die Liste von Listen, um die ursprüngliche Struktur wiederherzustellen
    loaded_features = list(zip(*loaded_features))
    loaded_features = [list(item) for item in loaded_features]

# Wenn die Features als Listen in einer Zelle gespeichert wurden (Szenario 2):
elif 'feature' in df_loaded.columns:
    loaded_features = [ast.literal_eval(item) for item in df_loaded['feature'].tolist()]

print("Daten erfolgreich aus CSV eingelesen:")
print("Geladene Labels (erste 5):", loaded_labels[:5])
print("Geladene Features (erste 1):", loaded_features[:1])


Daten erfolgreich aus CSV eingelesen:
Geladene Labels (erste 5): [0, 1, 8, 8, 9]
Geladene Features (erste 1): [[1.1573697328567505, 0.3032321631908417, 0.4962291121482849, 0.1067860126495361, -1.625678539276123, 0.4844562709331512, 0.6799285411834717, 0.4303563237190246, -1.5096973180770874, -1.3390100002288818, -0.6796276569366455, 0.913780689239502, 0.3822452127933502, 1.5332729816436768, 0.0255593322217464, -3.4735898971557617, 0.6965407133102417, 0.8458954691886902, 0.5373067855834961, -0.2713418304920196, 1.312721848487854, 4.076414585113525, -0.0822739154100418, 0.0968436002731323, -1.4931050539016724, -2.24968695640564, -0.820317804813385, -0.6016552448272705, 4.658712387084961, 0.5360066294670105, 0.9967613816261292, -0.3212272822856903, 1.1852117776870728, -1.3273863792419434, -2.471297264099121, -1.567915439605713, -0.9555495381355286, -1.0097880363464355, -0.481315404176712, 0.1386799365282058, 1.704442024230957, -1.3899831771850586, -1.121779203414917, 0.5226563811302185, -

## What are the classes per task?
Ich brauche eine übersetzung.  
Zusätzlich: Was ist mit den Seeds? Classenreihenfolge ist seed abhängig! Seed der runs berücksichtigen.

In [7]:
os.environ['CUDA_VISIBLE_DEVICES'] = "1"
args = parse_arguments()


seeds = [2000, 2001, 2002, 2003, 2004]
datasets = ["imagenetr", "cifar100", "cub", "dil_imagenetr", "imageneta", "cars", "omnibenchmark", "limited_domainnet"]


for seed in seeds:
    for dataset in datasets:
        args.seed = seed
        args.dataset = dataset
        args.device = "cpu"
        args = update_args(args)

        set_seed(args.seed)

        main2_classes_per_task_and_seed(args)
        break
    break

CIL


Iterate over train dataset: 24000it [00:00, 977436.92it/s]
Iterate over test dataset: 6000it [00:00, 1569920.40it/s]
Accumulating classes per task: 100%|██████████| 10/10 [02:14<00:00, 13.44s/it]

Classes per task saved in local_data/classes_per_task/CpT_imagenetr_2000.csv


## Which expert learned which class

In [84]:
wandb.login(key="8a88a8c49d1c2d31b8677fe0b8eb7d3e3a031f83")
api = wandb.Api()


def get_expert_distribution(run):
    if run.state != "finished":
        return None

    history = run.history()
    
    expert_distributions = dict()
    ft_tasks = [None] * 1000
    ft_buffer = run.config.get("moe_max_experts")
    for line in run.history().columns:
        if line.startswith("Expert") and line.endswith("learned task"):
            
            line_splited = line.split(" ")
            expert = int(line_splited[1])
            tasks = history[line].dropna().tolist()
            tasks = [int(task) for task in tasks]
            
            if expert not in expert_distributions:
                expert_distributions[expert] = list()
            expert_distributions[expert].extend(tasks) 

            
            for i in tasks:
                if i >= ft_buffer:
                    ft_tasks[i - ft_buffer] = expert    

    # cleaning ft_tasks
    ft_tasks = [i for i in ft_tasks if i is not None]

    return ft_tasks, expert_distributions

def get_sweep_data(runs, attributes_config=[], attributes_summary=[], include_run_id=True, class_similarity=False):
    sweep_data = []
    for run in runs:
        config = run.config
        summary = run.summary

        if summary.get("task_mean/acc") is not None and run.state == "finished":
            run_data = dict()
            if include_run_id:
                run_data["run_id"] = run.id
            # Add the config attributes to the run_data dictionary
            for attr in attributes_config:
                run_data[attr] = config.get(attr)

            # Add the summary attributes to the run_data dictionary
            for attr in attributes_summary:
                run_data[attr] = summary.get(attr)
            
            # average class similarity of learned classes per expert
            if class_similarity and run_data["dataset"] == DATASET:
                _, task_distribution = get_expert_distribution(run)
                print(f"Run {run.id} - Task Distribution: {task_distribution}")

                # Map classes per task
                classes_per_task = get_classes_per_task(run_data["dataset"], run_data["seed"])
                for expert_id, learned_tasks in task_distribution.items():
                    classes= []
                    for task in learned_tasks:
                        classes.extend(classes_per_task[task])
                    task_distribution[expert_id] = classes
            

                # Calculate inter-class similarity for each expert
                expert_similaritys = list()
                for expert_id, learned_classes in task_distribution.items():
                    if len(learned_classes) > 1:
                        label_mask = np.isin(loaded_labels, learned_classes)

                        # Verwende die Maske, um die entsprechenden Features und Labels auszuwählen
                        expert_features = loaded_features[label_mask]
                        expert_labels = loaded_labels[label_mask]
                        print(expert_labels)
                        print(f"Expert {expert_id} learned classes: {learned_classes}")
                        print(f"Shape: {expert_features.shape}")
                        similarity = calculate_intra_class_similarity_optimized(expert_features, expert_labels, similarity_metric='cosine')[1]
                        print(f"Expert {expert_id} - Intra-Class Similarity: {similarity:.4f}")
                        expert_similaritys.append(similarity)

                run_data["expert_similaritys"] = expert_similaritys
                average_similarity = np.mean(expert_similaritys) if expert_similaritys else 0.0       
                run_data["average_expert_similarity"] = average_similarity
                print(f"Average Expert Similarity: {average_similarity:.4f}")




            sweep_data.append(run_data)

    return sweep_data


def get_classes_per_task(dataset, seed):
    """
    Get the classes per task for a given dataset and seed.
    """
    csv_path = f"local_data/classes_per_task/CpT_{dataset}_{seed}.csv"
    if not os.path.exists(csv_path):
        print(f"CSV file {csv_path} does not exist.")
        return None

    df = pd.read_csv(csv_path)
    classes_per_task = {}
    for _, row in df.iterrows():
        task = row['task']
        class_label = row['class']
        if task not in classes_per_task:
            classes_per_task[task] = []
        classes_per_task[task].append(class_label)

    return classes_per_task

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: Appending key for api.wandb.ai to your netrc file: /export/home/0schindl/.netrc


## Average similarity per expert

In [105]:
DATASET = "omnibenchmark"

In [106]:
## Load data from file
csv_filename = f"./local_data/class_features/class_features_{DATASET}.csv"

# Lese die CSV-Datei ein
df_loaded = pd.read_csv(csv_filename)

# Extrahiere die Labels und Features
loaded_labels = df_loaded['label'].tolist()

# Wenn die Features als separate Spalten gespeichert wurden (Szenario 1):
if 'feature_0' in df_loaded.columns:
    loaded_features = []
    i = 0
    while f'feature_{i}' in df_loaded.columns:
        loaded_features.append(df_loaded[f'feature_{i}'].tolist())
        i += 1
    # Transponiere die Liste von Listen, um die ursprüngliche Struktur wiederherzustellen
    loaded_features = list(zip(*loaded_features))
    loaded_features = [list(item) for item in loaded_features]

# Wenn die Features als Listen in einer Zelle gespeichert wurden (Szenario 2):
elif 'feature' in df_loaded.columns:
    loaded_features = [ast.literal_eval(item) for item in df_loaded['feature'].tolist()]

print("Daten erfolgreich aus CSV eingelesen:")
print("Geladene Labels (erste 5):", loaded_labels[:5])
print("Geladene Features (erste 1):", loaded_features[:1])


Daten erfolgreich aus CSV eingelesen:
Geladene Labels (erste 5): [0, 0, 0, 0, 0]
Geladene Features (erste 1): [[1.9689552783966064, 0.0870765522122383, -0.0113297663629055, 0.6839339137077332, 0.1754369139671325, 0.9810996055603028, -0.4749040901660919, 1.8897793292999268, -0.4923095107078552, -0.4473477303981781, -1.486798882484436, 0.8747390508651733, -0.0753733217716217, -0.6563200950622559, 0.718726634979248, -0.4465139806270599, 1.195293664932251, 0.6493381857872009, 0.6927939653396606, -0.5141491293907166, 0.8577017188072205, -0.3272809982299804, -1.0333484411239624, 0.4337775409221649, -1.115162372589111, -0.0877170041203498, 1.3080565929412842, 0.7482463121414185, -0.8268628716468811, -0.1584079265594482, -0.5005767941474915, -1.4997069835662842, 2.248469591140747, -1.0296906232833862, -0.1423937827348709, -0.4433310627937317, -0.1288578808307647, 0.3147761523723602, -0.3346352875232696, -0.6630178689956665, -0.0858169347047805, -0.3058180212974548, -0.4738557934761047, -0.5707

In [107]:

_42_adapter_performance = [
    "belaschindler-university-hamburg/0schindl-LayUp_sweeps_question1_results/6kim8tiu", # DIL
    "belaschindler-university-hamburg/0schindl-LayUp_sweeps_question1_results/jdpa9z1x", # Cars
    "belaschindler-university-hamburg/0schindl-LayUp_sweeps_question1_results/cjddpel4", # Imagenet-a
    "belaschindler-university-hamburg/0schindl-LayUp_sweeps_question1_results/hxigp6ck", # Imagenet-r
    "belaschindler-university-hamburg/0schindl-LayUp_sweeps_question1_results/p7zmthx9", # CIL
    ]
datsets_in_CIL = ["cifar100", "cub", "vtab", "omnibenchmark"]

loaded_labels = np.array(loaded_labels)
loaded_features = np.array(loaded_features)

table_421 = []
for i, s in enumerate(_42_adapter_performance):
    sweep = api.sweep(s)
    runs = sweep.runs

    attributes_config = ["dataset", "selection_method", "seed"]
    attributes_summary = ["task_mean/acc"]

    data = get_sweep_data(runs, attributes_config, attributes_summary, class_similarity=True)
    for e in data:
        if e["dataset"] == DATASET:
            if i == 4:
                # CIL
                if e["dataset"] in datsets_in_CIL:
                    table_421.append(e)
            else:
                table_421.append(e)
    


df_sweep = pd.DataFrame(table_421)

Run 6dwyk5lm - Task Distribution: {4: [4, 5, 7], 0: [0, 6], 2: [2, 8, 9], 1: [1], 3: [3]}
[  2   2   2 ... 299 299 299]
Expert 4 learned classes: [2, 13, 49, 53, 61, 65, 66, 73, 75, 77, 81, 84, 94, 100, 102, 113, 128, 145, 154, 171, 199, 216, 224, 229, 243, 258, 277, 285, 288, 290, 6, 26, 29, 30, 39, 51, 103, 105, 110, 116, 117, 121, 142, 152, 172, 181, 182, 190, 198, 211, 215, 218, 221, 223, 235, 250, 257, 265, 279, 294, 10, 12, 22, 23, 55, 69, 76, 83, 93, 101, 140, 160, 161, 167, 168, 170, 176, 209, 233, 236, 244, 263, 269, 271, 280, 286, 289, 292, 293, 299]
Shape: (26911, 768)


Calculating Intra-Class Similarity: 100%|██████████| 90/90 [00:00<00:00, 260.36class/s]


Expert 4 - Intra-Class Similarity: 0.5415
[  5   5   5 ... 297 297 297]
Expert 0 learned classes: [5, 17, 38, 50, 58, 67, 70, 74, 89, 99, 106, 114, 118, 119, 122, 130, 136, 137, 151, 166, 177, 197, 205, 210, 219, 220, 248, 252, 270, 297, 8, 16, 31, 34, 36, 42, 48, 60, 82, 86, 92, 104, 112, 120, 126, 138, 139, 143, 148, 158, 165, 185, 200, 201, 202, 228, 241, 275, 287, 296]
Shape: (18264, 768)


Calculating Intra-Class Similarity: 100%|██████████| 60/60 [00:00<00:00, 243.82class/s]


Expert 0 - Intra-Class Similarity: 0.5497
[  1   1   1 ... 298 298 298]
Expert 2 learned classes: [32, 35, 57, 64, 68, 80, 95, 111, 115, 125, 131, 132, 146, 150, 155, 179, 187, 188, 193, 196, 212, 231, 232, 238, 240, 253, 262, 266, 267, 273, 7, 9, 21, 24, 37, 40, 44, 62, 78, 79, 85, 96, 97, 107, 134, 135, 141, 159, 174, 184, 208, 226, 242, 246, 256, 259, 260, 276, 278, 291, 1, 4, 14, 15, 28, 41, 43, 52, 54, 63, 71, 91, 109, 123, 124, 164, 178, 183, 192, 194, 206, 207, 214, 217, 247, 264, 268, 272, 284, 298]
Shape: (26990, 768)


Calculating Intra-Class Similarity: 100%|██████████| 90/90 [00:00<00:00, 258.59class/s]


Expert 2 - Intra-Class Similarity: 0.5378
[  3   3   3 ... 282 282 282]
Expert 1 learned classes: [3, 19, 33, 45, 47, 59, 72, 90, 98, 127, 129, 156, 157, 169, 173, 175, 180, 195, 203, 222, 227, 230, 234, 237, 239, 249, 254, 274, 281, 282]
Shape: (8709, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 272.36class/s]


Expert 1 - Intra-Class Similarity: 0.5009
[  0   0   0 ... 295 295 295]
Expert 3 learned classes: [0, 11, 18, 20, 25, 27, 46, 56, 87, 88, 108, 133, 144, 147, 149, 153, 162, 163, 186, 189, 191, 204, 213, 225, 245, 251, 255, 261, 283, 295]
Shape: (8823, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 246.19class/s]


Expert 3 - Intra-Class Similarity: 0.5251
Average Expert Similarity: 0.5310
Run 7z247qq2 - Task Distribution: {0: [0], 1: [1, 8], 3: [3], 2: [2, 7], 4: [4, 5, 6, 9]}
[  5   5   5 ... 298 298 298]
Expert 0 learned classes: [5, 22, 24, 33, 43, 50, 52, 53, 73, 88, 99, 115, 121, 125, 130, 148, 161, 165, 170, 182, 223, 231, 234, 247, 252, 254, 268, 277, 281, 298]
Shape: (8582, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 216.92class/s]

Expert 0 - Intra-Class Similarity: 0.4892


[  3   3   3 ... 299 299 299]
Expert 1 learned classes: [10, 13, 26, 34, 44, 64, 68, 71, 89, 90, 101, 107, 129, 137, 149, 155, 171, 173, 176, 190, 200, 218, 227, 242, 245, 249, 278, 282, 293, 297, 3, 23, 51, 54, 58, 60, 75, 78, 95, 106, 120, 136, 139, 156, 163, 166, 168, 183, 198, 202, 207, 215, 219, 222, 232, 251, 253, 258, 292, 299]
Shape: (18139, 768)


Calculating Intra-Class Similarity: 100%|██████████| 60/60 [00:00<00:00, 262.10class/s]


Expert 1 - Intra-Class Similarity: 0.5505
[ 11  11  11 ... 296 296 296]
Expert 3 learned classes: [11, 21, 31, 39, 77, 81, 103, 109, 112, 122, 124, 141, 142, 147, 151, 153, 159, 174, 179, 196, 197, 208, 228, 244, 257, 273, 283, 288, 295, 296]
Shape: (8916, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 265.16class/s]


Expert 3 - Intra-Class Similarity: 0.5361
[  8   8   8 ... 290 290 290]
Expert 2 learned classes: [8, 9, 12, 18, 36, 41, 57, 80, 82, 92, 94, 117, 123, 127, 140, 158, 172, 187, 188, 199, 210, 213, 217, 225, 233, 238, 243, 264, 276, 286, 17, 32, 35, 55, 59, 63, 74, 93, 100, 102, 105, 143, 144, 146, 157, 164, 177, 181, 195, 201, 203, 211, 216, 229, 230, 236, 255, 265, 289, 290]
Shape: (18213, 768)


Calculating Intra-Class Similarity: 100%|██████████| 60/60 [00:00<00:00, 251.79class/s]


Expert 2 - Intra-Class Similarity: 0.5626
[  0   0   0 ... 294 294 294]
Expert 4 learned classes: [4, 27, 28, 30, 38, 45, 48, 49, 69, 72, 83, 87, 108, 116, 118, 133, 135, 138, 162, 180, 185, 206, 214, 248, 259, 261, 262, 263, 275, 287, 6, 15, 16, 19, 29, 56, 62, 67, 76, 79, 84, 97, 98, 104, 111, 119, 152, 175, 184, 191, 204, 205, 209, 221, 235, 237, 240, 246, 270, 291, 0, 1, 2, 37, 40, 46, 47, 61, 65, 66, 85, 91, 110, 114, 131, 132, 134, 150, 154, 167, 169, 189, 194, 212, 224, 241, 271, 279, 280, 284, 7, 14, 20, 25, 42, 70, 86, 96, 113, 126, 128, 145, 160, 178, 186, 192, 193, 220, 226, 239, 250, 256, 260, 266, 267, 269, 272, 274, 285, 294]
Shape: (35847, 768)


Calculating Intra-Class Similarity: 100%|██████████| 120/120 [00:00<00:00, 258.18class/s]


Expert 4 - Intra-Class Similarity: 0.5279
Average Expert Similarity: 0.5333
Run 5b5h2uzl - Task Distribution: {3: [3], 0: [0, 6], 4: [4], 1: [1], 2: [2, 5, 7, 8, 9]}
[  1   1   1 ... 285 285 285]
Expert 3 learned classes: [1, 4, 14, 16, 27, 29, 45, 51, 61, 77, 97, 105, 131, 141, 143, 167, 169, 170, 177, 181, 198, 203, 211, 223, 239, 247, 256, 265, 282, 285]
Shape: (8902, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 256.33class/s]

Expert 3 - Intra-Class Similarity: 0.4967


[ 34  34  34 ... 299 299 299]
Expert 0 learned classes: [34, 48, 54, 60, 64, 74, 79, 99, 100, 107, 128, 136, 144, 162, 165, 188, 189, 197, 213, 219, 224, 226, 237, 240, 246, 251, 259, 263, 294, 299, 53, 66, 75, 89, 90, 93, 96, 102, 110, 118, 120, 124, 148, 172, 174, 179, 195, 200, 201, 206, 207, 208, 216, 230, 235, 249, 250, 274, 283, 290]
Shape: (18092, 768)


Calculating Intra-Class Similarity: 100%|██████████| 60/60 [00:00<00:00, 259.28class/s]


Expert 0 - Intra-Class Similarity: 0.5529
[  2   2   2 ... 293 293 293]
Expert 4 learned classes: [2, 8, 24, 35, 36, 40, 41, 44, 57, 59, 70, 76, 83, 112, 130, 133, 142, 146, 147, 156, 158, 168, 194, 205, 221, 225, 253, 257, 273, 293]
Shape: (9181, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 253.15class/s]


Expert 4 - Intra-Class Similarity: 0.5571
[  9   9   9 ... 297 297 297]
Expert 1 learned classes: [9, 17, 19, 25, 28, 31, 32, 38, 39, 56, 62, 63, 68, 72, 98, 116, 125, 161, 182, 190, 193, 204, 220, 242, 243, 248, 270, 272, 280, 297]
Shape: (8873, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 265.90class/s]


Expert 1 - Intra-Class Similarity: 0.5236
[  0   0   0 ... 298 298 298]
Expert 2 learned classes: [13, 20, 23, 30, 55, 65, 71, 81, 86, 87, 91, 122, 138, 149, 150, 153, 164, 166, 180, 187, 191, 214, 217, 227, 234, 236, 241, 255, 289, 295, 18, 37, 46, 50, 52, 58, 69, 84, 92, 108, 113, 121, 123, 135, 145, 163, 171, 178, 186, 196, 202, 209, 228, 233, 252, 254, 261, 266, 281, 296, 0, 3, 15, 22, 67, 82, 88, 95, 101, 114, 119, 126, 140, 154, 155, 160, 173, 175, 218, 232, 244, 245, 260, 268, 271, 275, 284, 286, 292, 298, 10, 11, 12, 21, 42, 43, 47, 49, 73, 78, 80, 94, 103, 104, 115, 127, 132, 134, 137, 151, 159, 176, 183, 184, 185, 199, 264, 267, 269, 279, 5, 6, 7, 26, 33, 85, 106, 109, 111, 117, 129, 139, 152, 157, 192, 210, 212, 215, 222, 229, 231, 238, 258, 262, 276, 277, 278, 287, 288, 291]
Shape: (44649, 768)


Calculating Intra-Class Similarity: 100%|██████████| 150/150 [00:00<00:00, 258.07class/s]


Expert 2 - Intra-Class Similarity: 0.5360
Average Expert Similarity: 0.5333
Run fs3coo40 - Task Distribution: {0: [0, 5, 8, 9], 3: [3, 6, 7], 4: [4], 2: [2], 1: [1]}
[  0   0   0 ... 299 299 299]
Expert 0 learned classes: [10, 14, 23, 30, 50, 76, 87, 90, 92, 99, 111, 128, 138, 147, 166, 175, 176, 179, 187, 194, 202, 216, 224, 228, 232, 258, 262, 263, 295, 299, 3, 21, 25, 26, 32, 38, 44, 45, 54, 60, 66, 75, 83, 86, 112, 115, 142, 157, 168, 196, 206, 219, 252, 259, 271, 272, 288, 293, 296, 298, 0, 20, 28, 29, 41, 47, 58, 62, 71, 82, 88, 93, 103, 107, 126, 133, 139, 140, 143, 156, 199, 211, 223, 233, 235, 239, 241, 246, 249, 268, 7, 9, 22, 56, 59, 79, 81, 104, 125, 127, 130, 153, 154, 160, 165, 172, 185, 189, 195, 201, 203, 214, 218, 247, 248, 257, 269, 273, 280, 294]
Shape: (36452, 768)


Calculating Intra-Class Similarity: 100%|██████████| 120/120 [00:00<00:00, 255.72class/s]


Expert 0 - Intra-Class Similarity: 0.5428
[  1   1   1 ... 297 297 297]
Expert 3 learned classes: [4, 12, 31, 55, 61, 65, 69, 70, 77, 91, 100, 101, 119, 121, 132, 136, 150, 151, 163, 181, 184, 204, 236, 243, 256, 275, 276, 277, 283, 297, 1, 13, 35, 68, 98, 102, 109, 116, 122, 123, 134, 141, 144, 155, 167, 171, 174, 178, 188, 208, 209, 217, 229, 234, 242, 253, 255, 265, 267, 274, 19, 37, 40, 46, 51, 53, 63, 64, 73, 113, 118, 129, 146, 148, 158, 169, 186, 198, 205, 212, 220, 230, 237, 238, 251, 264, 279, 282, 290, 292]
Shape: (26665, 768)


Calculating Intra-Class Similarity: 100%|██████████| 90/90 [00:00<00:00, 262.36class/s]


Expert 3 - Intra-Class Similarity: 0.5117
[  6   6   6 ... 289 289 289]
Expert 4 learned classes: [6, 16, 27, 33, 43, 49, 52, 57, 74, 84, 85, 95, 96, 105, 106, 117, 120, 162, 164, 170, 190, 192, 197, 221, 240, 260, 261, 270, 278, 289]
Shape: (9021, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 257.41class/s]


Expert 4 - Intra-Class Similarity: 0.5876
[  2   2   2 ... 287 287 287]
Expert 2 learned classes: [2, 11, 15, 17, 39, 78, 80, 89, 94, 97, 108, 131, 137, 149, 159, 180, 182, 183, 191, 193, 200, 210, 222, 225, 250, 281, 284, 285, 286, 287]
Shape: (8993, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 263.62class/s]


Expert 2 - Intra-Class Similarity: 0.5503
[  5   5   5 ... 291 291 291]
Expert 1 learned classes: [5, 8, 18, 24, 34, 36, 42, 48, 67, 72, 110, 114, 124, 135, 145, 152, 161, 173, 177, 207, 213, 215, 226, 227, 231, 244, 245, 254, 266, 291]
Shape: (8566, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 275.26class/s]


Expert 1 - Intra-Class Similarity: 0.5190
Average Expert Similarity: 0.5423
Run 26v6mgz7 - Task Distribution: {3: [3], 4: [4], 1: [1, 5, 6], 2: [2, 7, 8, 9], 0: [0]}
[  3   3   3 ... 292 292 292]
Expert 3 learned classes: [3, 35, 47, 53, 76, 87, 108, 110, 115, 118, 122, 131, 132, 144, 148, 153, 154, 163, 170, 191, 226, 228, 229, 236, 239, 241, 277, 288, 289, 292]
Shape: (8769, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 264.54class/s]


Expert 3 - Intra-Class Similarity: 0.5484
[  7   7   7 ... 299 299 299]
Expert 4 learned classes: [7, 10, 12, 13, 14, 40, 50, 57, 69, 79, 90, 106, 107, 116, 135, 142, 171, 184, 198, 203, 205, 210, 222, 223, 238, 262, 269, 270, 281, 299]
Shape: (9158, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 252.93class/s]


Expert 4 - Intra-Class Similarity: 0.5375
[  0   0   0 ... 298 298 298]
Expert 1 learned classes: [6, 11, 15, 19, 20, 23, 27, 43, 51, 54, 80, 94, 100, 117, 119, 136, 179, 180, 183, 188, 196, 216, 217, 220, 232, 246, 251, 267, 279, 298, 17, 22, 37, 44, 61, 99, 104, 114, 138, 140, 160, 169, 173, 195, 199, 202, 207, 214, 218, 219, 233, 237, 247, 271, 274, 275, 278, 280, 282, 297, 0, 5, 45, 52, 58, 59, 63, 67, 68, 71, 74, 83, 93, 123, 130, 149, 151, 161, 162, 164, 172, 189, 190, 201, 213, 245, 263, 268, 276, 291]
Shape: (27151, 768)


Calculating Intra-Class Similarity: 100%|██████████| 90/90 [00:00<00:00, 253.11class/s]


Expert 1 - Intra-Class Similarity: 0.5414
[  1   1   1 ... 296 296 296]
Expert 2 learned classes: [1, 18, 21, 36, 38, 55, 70, 77, 78, 81, 85, 88, 91, 92, 112, 113, 145, 158, 167, 182, 225, 227, 234, 243, 248, 258, 259, 261, 264, 294, 2, 25, 34, 64, 65, 75, 89, 95, 96, 109, 139, 152, 166, 176, 181, 185, 204, 212, 221, 231, 242, 254, 266, 272, 273, 284, 287, 290, 293, 296, 24, 28, 29, 31, 72, 73, 82, 84, 102, 103, 111, 124, 126, 127, 133, 134, 137, 143, 155, 174, 206, 211, 230, 235, 240, 249, 265, 283, 285, 286, 4, 16, 26, 39, 42, 48, 60, 62, 86, 101, 105, 120, 121, 128, 129, 141, 146, 147, 175, 177, 186, 197, 208, 224, 250, 253, 256, 257, 260, 295]
Shape: (35984, 768)


Calculating Intra-Class Similarity: 100%|██████████| 120/120 [00:00<00:00, 256.68class/s]


Expert 2 - Intra-Class Similarity: 0.5348
[  8   8   8 ... 255 255 255]
Expert 0 learned classes: [8, 9, 30, 32, 33, 41, 46, 49, 56, 66, 97, 98, 125, 150, 156, 157, 159, 165, 168, 178, 187, 192, 193, 194, 200, 209, 215, 244, 252, 255]
Shape: (8635, 768)


Calculating Intra-Class Similarity: 100%|██████████| 30/30 [00:00<00:00, 270.10class/s]

Expert 0 - Intra-Class Similarity: 0.5143
Average Expert Similarity: 0.5353


In [108]:
data = df_sweep
print(data)

     run_id        dataset selection_method  seed  task_mean/acc  \
0  6dwyk5lm  omnibenchmark       inv_ws_div  2000       0.573589   
1  7z247qq2  omnibenchmark       inv_ws_div  2004       0.581586   
2  5b5h2uzl  omnibenchmark       inv_ws_div  2003       0.597296   
3  fs3coo40  omnibenchmark       inv_ws_div  2002       0.585258   
4  26v6mgz7  omnibenchmark       inv_ws_div  2001       0.591332   

                                  expert_similaritys  \
0  [0.5415225760440951, 0.5496745796154197, 0.537...   
1  [0.4892236990547385, 0.5505489179248585, 0.536...   
2  [0.4966600350786431, 0.5529010815099288, 0.557...   
3  [0.5428289046490888, 0.5117119500755918, 0.587...   
4  [0.5483583799959261, 0.5374967049866456, 0.541...   

   average_expert_similarity  
0                   0.531002  
1                   0.533287  
2                   0.533252  
3                   0.542283  
4                   0.535253  


### CARS

   average_expert_similarity  
0                   0.673711  
1                   0.666535  
2                   0.668896  
3                   0.672675  
4                   0.666224  

### CIFAR
0                   0.477616  
1                   0.471247  
2                   0.473173  
3                   0.477304  
4                   0.474774  


### CUB
0                   0.629865  
1                   0.636929  
2                   0.630805  
3                   0.623717  
4                   0.638992  

### IN-A
0                   0.179068  
1                   0.181273  
2                   0.184007  
3                   0.185891  
4                   0.184179  

### IN-R
0                   0.268898  
1                   0.274224  
2                   0.276426  
3                   0.280183  
4                   0.278103  

## OB
0                   0.531002  
1                   0.533287  
2                   0.533252  
3                   0.542283  
4                   0.535253  



In [ ]:

                # Calculate inter-class similarity for each expert
                expert_similaritys = list()
                for expert_id, learned_classes in task_distribution.items():
                    if len(learned_classes) > 1:
                        label_mask = np.isin(loaded_labels, learned_classes)

                        # Verwende die Maske, um die entsprechenden Features und Labels auszuwählen
                        expert_features = loaded_features[label_mask]
                        expert_labels = loaded_labels[label_mask]
                        print(f"Expert {expert_id} learned classes: {learned_classes}")
                        print(f"Shape: {expert_features.shape}")
                        similarity = calculate_inter_class_similarity_vectorized(expert_features, expert_labels)
                        print(f"Expert {expert_id} - Inter-Class Similarity: {similarity:.4f}")
                        expert_similaritys.append(similarity)

                run_data["expert_similaritys"] = expert_similaritys
                average_similarity = np.mean(expert_similaritys) if expert_similaritys else 0.0       
                run_data["average_expert_similarity"] = average_similarity
                print(f"Average Expert Similarity: {average_similarity:.4f}")
                print(task_distribution)


                # inter-class similarity to all classes except the one where their expert only learned this one class
                lonly_learned_classes = []
                for expert_id, learned_classes in task_distribution.items():
                    if len(learned_classes) == 1:
                        lonly_learned_classes.append(learned_classes[0])
                print(f"Classes where expert only learned this class: {lonly_learned_classes}")

                # Calculate inter-class similarity to all other classes
                other_classes = [label for label in loaded_labels if label != lonly_learned_classes]
                label_mask = np.isin(loaded_labels, other_classes)

                expert_features = loaded_features[label_mask]
                expert_labels = loaded_labels[label_mask]
                similarity = calculate_inter_class_similarity_vectorized(expert_features, expert_labels)
                run_data["filtered_dataset_similarity"] = similarity
                print(f"Dataset similarity: {similarity:.4f}")


